Calculator [STEP3] Mason, Blue

calledBlu opened the pull request:
안녕하세요 라자냐! @wonhee009
팀 🔮MABLE🔮 의 Mason, Blue 입니다 😊
STEP 3 요구사항 중 미완료 항목 부분은 PR 마감 권장기간에 맞추기 위해 후순위로 생각하였습니다.
해당 부분은 PR 전송 후 틈틈이 리팩토링 진행하도록 하겠습니다!
리뷰 잘 부탁드립니다! 감사합니다 💙

# ✅ 요구사항 & 체크리스트

- 연산자, 숫자 입력 검증
    - [x]  연산자, 숫자 입력에 대한 검증
    - [x]  현재 숫자입력이 없는 상태인 0에서는 연산자를 반복해서 누르더라도 연산이 이뤄지지 않습니다
    - [x]  현재 숫자입력이 없는 상태인 0에서는 연산자의 종류만 변경합니다
    - [x]  숫자입력 중에 연산자(÷, ×, -, +)를 누르게 되면 숫자입력을 중지하고 다음 숫자를 입력받습니다
    - ⁺⁄₋ 버튼 누를 시 처리
    - [x]  현재 입력한 숫자의 부호를 변환합니다
    - [x]  입력된 숫자가 0인경우 부호를 표시하지 않고 변경하지도 않습니다
- 계산 결과 나타내기
    - [x]  = 버튼을 누르면 입력된 연산을 한 번에 수행합니다
    - [x]  = 버튼을 눌러 연산을 마친 후 다시 =을 눌러도 이전 연산을 다시 연산하지 않습니다
    - [x]  = 버튼을 누른 후 연산자를 누르면 결과값으로 계산할 수 있도록 구현
    - [x]  = 버튼을 누른 후 숫자를 누르면 초기화
    - [x]  0으로 나누기에 대해서는 결과값을 NaN으로 표기합니다
- 사용자 반응
    - [x]  스택뷰와 레이블을 활용하여 계산 내역을 표시해줍니다
    - [x]  계산내역이 상단 공간을 넘어 이어지는 경우, 사용자에게 제대로 보일 수 있도록 새로 추가될 때 최하단으로 자동 스크롤 할 수 있도록 합니다
- 숫자 제약조건
    - [x]  숫자는 3자리마다 쉼표(,)를 표기해줍니다
    - [x]  Double 연산 오차 수정 (부동소수점 오류…)
    - [x]  사용자에게 표시하는 숫자는 뒤에 0000 등 불필요한 숫자가 나타나지 않도록 합니다
        - 예) 0.123000000 → 0.123 / 1.22340000 → 1.2234
- 기타
    - [x]  매직넘버, 매직스트링 상수로 관리하게 처리

> 🙅‍♀️ 미완료 항목
- 숫자 제약조건
    - [ ]  숫자는 최대 20자리까지만 표현합니다 (잘리는 숫자는 반올림합니다)

# ✏️ 학습 키워드

- StackView
- ScrollView
- layoutIfNeeded()
- inout

# **⁉️ 고민과 해결**

계산기가 수행하는 기본적인 동작은 구현 완료 하였으나, 아직 세부적인 디테일을 미완료한 상태입니다.

지금까지 작성한 코드 중에서 네이밍, 입력 검증 로직과 같은 예외처리를 중점적으로 개선해야 할 부분들이 있다면 피드백 부탁드리겠습니다! 😄

또한, 실제 앱 작동 시 위험하게 생각되는 요소들이 있다면 의견 부탁드립니다!

calledBlu committed:
feat: 연산자 및 피연산자 레이블 VC에 연결

qwerty3345 committed:
feat: 숫자 operand 버튼 입력 시 검증 로직 추가

calledBlu committed:
feat: tappedOperatorButton() 메서드 내부 구현

- 연산자를 누르는 경우 앞서 입력된 연산자와 입력된 피연산자가  scrollView 내의 Stack에 추가되도록 구현

qwerty3345 committed:
refactor: tappedOperatorButton 메서드 로직 함수 분리 리팩터링

calledBlu committed:
feat: tappedOperatorButton() 메서드 구현 및 tappedACButton() 메서드 구현

- 연산자 버튼을 누르는 경우 연산자 입력 검증 수행
- AC 버튼을 눌렀을 때 historyStackView 내부의 view들을 삭제하고 formulaString, operandLabel, operatorLabel 초기화

qwerty3345 committed:
feat: = 버튼 누를 시 결과 계산 로직 구현

-  = 버튼을 누른 후 연산자를 누르면 결과값으로 계산할 수 있도록 구현
-  = 버튼을 누른 후 숫자를 누르면 초기화

qwerty3345 committed:
fix: NaN 상태일 때의 연산 예외 처리

calledBlu committed:
chore: 실행과 관련 없는 print() 메서드 삭제

qwerty3345 committed:
refactor: 함수 재사용 분리

wonhee009 commented on Calculator/Calculator/Extensions/String+Extensions.swift:
```diff
@@ -11,4 +11,9 @@ extension String {
         return components(separatedBy: String(target))
     }
+
+    var isNumber: Bool {
+        let digitsCharacters = CharacterSet(charactersIn: "0123456789")
+        return CharacterSet(charactersIn: self).isSubset(of: digitsCharacters)
+    }
```
문자열이 숫자인지 확인하는 코드인거 같은데요. Int로 변환이 가능한지 아닌지 확인하는 방식으로는 안 되나요?

calledBlu replied:
기존에는 숫자인지를 확인하려고 하였으나, 
문자열이 숫자로만 이루어져있는 것을 확인하기보다는 `.` 혹은 `0`으로 이루어져 있는지를 확인하는 방향으로 수정하였습니다.

> 해당 커밋: ([df9b4bc](https://github.com/tasty-code/ios-calculator/pull/28/commits/df9b4bc8b1dd2771196508f1cc58026fed27e6fe))

wonhee009 commented on Calculator/Calculator/Controller/CalculatorViewController.swift:
```diff
@@ -0,0 +1,154 @@
+            isCalculated = false
+        }
+
+        guard var operand = operandLabel.text else { return }
+        guard let inputOperand = sender.titleLabel?.text else { return }
```
```suggestion
        guard var operand = operandLabel.text,
                   let inputOperand = sender.titleLabel?.text else { return }
```
이렇게 가능하지 않나요?

qwerty3345 replied:
축약 하였습니다 :)
> 해당 커밋: d190508

wonhee009 commented on Calculator/Calculator/Controller/CalculatorViewController.swift:
```diff
@@ -0,0 +1,154 @@
+        guard var operand = operandLabel.text else { return }
+        guard let inputOperand = sender.titleLabel?.text else { return }
+
+        if operand.contains(".") && inputOperand == "." {
+            return
+        }
+
+        if (operand == "0" || operand == "-0") && (inputOperand == "0" || inputOperand == "00") {
+            return
+        }
+
+        if (operand == "0" || operand == "-0") && 1...9 ~= (Int(inputOperand) ?? 0) {
+            operand.remove(at: operand.firstIndex(of: "0") ?? operand.startIndex)
+            operandLabel.text = operand + inputOperand
+            return
+        }
```
이런 부분들은 입력된 operand가 화면에 표시될 수 있는지 검증하는 코드라고 생각되는데요. 메서드로 분리하는건 어떤가요?

calledBlu replied:
라자냐가 추천해주신 방법대로 메서드로 분리해 보려고 하였습니다.
tappedOperandButton() 내의 검증 코드를 메서드로 분리하면서 tappedOperatorButton() 내의 검증 코드도 메서드로 분리해 보았습니다!
> tappedOperandButton() 검증 코드 메서드 분리: ([57bdabe](https://github.com/tasty-code/ios-calculator/pull/28/commits/57bdabed93c9c21ea4b169750d9c7f6570f0c32d))

> tappedOperatorButton() 검증 코드 메서드 분리: ([c6a94f3](https://github.com/tasty-code/ios-calculator/pull/28/commits/c6a94f39437eb930bcb82263f1005a2d2770a8ff))

wonhee009 commented on Calculator/Calculator/Controller/CalculatorViewController.swift:
```diff
@@ -0,0 +1,154 @@
+
+        if (operand == "0" || operand == "-0") && 1...9 ~= (Int(inputOperand) ?? 0) {
+            operand.remove(at: operand.firstIndex(of: "0") ?? operand.startIndex)
+            operandLabel.text = operand + inputOperand
+            return
```
아래에서 동일한 코드 `operandLabel.text = operand + inputOperand`가 있는거 같습니다.

qwerty3345 replied:
반복되는 로직을 처리하면서 함께 처리 해 주었습니다 :)

> 해당 커밋: 57bdabed93c9c21ea4b169750d9c7f6570f0c32d

wonhee009 commented on Calculator/Calculator/Controller/CalculatorViewController.swift:
```diff
@@ -0,0 +1,154 @@
+        updateFormulaString()
+        addHistoryLabels()
+
+        var fomula = ExpressionParser.parse(from: formulaString)
```
```suggestion
        var formula = ExpressionParser.parse(from: formulaString)
```

qwerty3345 replied:
오타 지적 감사합니다..! 꼼꼼히 확인하겠습니다 😭

wonhee009 commented on Calculator/Calculator/Controller/CalculatorViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    private func resetLabels() {
+        operandLabel.text = "0"
```
"0"이 반복적으로 사용되는거 같은데요. 상수로 정의하면 어떤가요?

qwerty3345 replied:
반복적으로 사용되는 string, character 값들을 extension의 상수로 정의하였습니다 :)

> 해당 커밋: 7634b4bed3140eb56e1be2468fb2b73a82c07a72

wonhee009 reviewed (changes requested):
수고하셨습니다:)
코멘트 남겨두었으니 확인 부탁드려요.

calledBlu committed:
fix: 연산 수행 로직 수정

- formulaString의 초기값을 0 ->  으로 수정
- 초기화 된 상태인지를 확인하여, 해당 상태일 때 연산자 버튼을 누르면 stack에 0이 추가되도록 수정
- 숫자가 0일 때는 부호가 적용되지 않도록 수정
- 위의 로직 수정에 따른 if 조건문 수정

qwerty3345 committed:
refactor: 여러 guard문을 한 guard문으로 축약

calledBlu committed:
refactor: tappedOperandButton() 메서드 내의 operand 검증 코드를 메서드로 분리

qwerty3345 committed:
refactor: tappedOperatorButton() 메서드 내의 operator 검증 코드를 메서드로 분리

calledBlu committed:
refactor: tappedOperandButton() 메서드 내의 중복되는 코드 수정

- isValid(operand: withInputOperand:) 메서드의 operand 파라미터를 inout으로 수정

qwerty3345 committed:
refactor: 반복적으로 사용되는 string, character 값들 extension의 상수로 정의

calledBlu committed:
refactor: isEmpty 메서드 사용으로 인해 불필요해진 초기화 상태를 확인하는 프로퍼티 및 로직 제거

qwerty3345 committed:
feat: 세자리마다 쉼표(,)를 주는 로직 추가

calledBlu committed:
refactor: .과 0으로만 구성된 피연산자의 경우 연산자를 누르면 stackView에 반영되지 않도록 수정

qwerty3345 committed:
feat: 계산 식 입력 시, scrollView 최하단으로 이동하도록 구현

qwerty3345 committed:
refactor: formatter 를 매번 생성하는 것이 아닌, 초기값으로 생성시켜놓고 사용하도록 변경

calledBlu committed:
Update README.md

wonhee009 commented on Calculator/Calculator/Extensions/String+Extensions.swift:
```diff
@@ -12,6 +12,10 @@ extension String {
         return components(separatedBy: String(target))
     }
 
+    var isZero: Bool {
+        allSatisfy ({ $0 == .dot || $0 == .zero })
+    }
```
String extension에 정의하면 모든 String 타입에서 해당 computed property를 사용할 수 있을텐데요.
모든 String 타입이 `isZero`임을 알아야하나요?

wonhee009 commented on Calculator/Calculator/Extensions/String+Extensions.swift:
```diff
@@ -11,4 +11,15 @@ extension String {
+        allSatisfy ({ $0 == .dot || $0 == .zero })
+    }
+
+    static let zero = "0"
+    static let dot = "."
+    static let NaN = "NaN"
+    static let dash = "-"
+    static let doubleZero = "00"
+    static let zeroDot = "0."
```
이 친구들도 String에 있어야하는 값인가요?

wonhee009 commented on Calculator/Calculator/Extensions/Character+Extensions.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+extension Character {
+    static let zero = Character("0")
+    static let dot = Character(".")
+    static let dash = Character("-")
+}
```
이 친구들 또한 마찬가지입니다.
